File: desktop/LegionControl.Desktop/Contract/AgentCapabilities.cs
namespace LegionControl.Desktop.Contract;

/// What the agent on the far side can be asked to do.
///
/// One number settles it. Every reply from a 3.x agent carries `"contract": 3`, and a client gates
/// every v3 feature on `contract >= 3`; absence means a 2.x agent and only the baseline verbs
/// exist there. That is deliberately not a per-verb capability list: an agent either implements
/// contract 3 or it does not, and a partial one would be a bug rather than a configuration.
///
/// Nothing here guesses. A verb the far side has not shown it understands is never sent, and a
/// control that needs one says what it needs instead of failing when pressed.
public sealed record AgentCapabilities
{
    /// The `contract` number from the last reply. Zero until something has answered.
    public int ContractVersion { get; init; }
    public string? AgentVersion { get; init; }

    public static readonly AgentCapabilities None = new();

    public static AgentCapabilities Of(AgentStatus status) => new()
    {
        ContractVersion = status.ContractVersion,
        AgentVersion = status.AgentVersion,
    };

    public static AgentCapabilities Of(AgentEnvelope envelope) => new()
    {
        ContractVersion = envelope.ContractVersion ?? 0,
        AgentVersion = envelope.AgentVersion,
    };

    /// Whether every v3 command and flag this app knows is safe to send.
    public bool SpeaksV3 => ContractVersion >= AgentContract.RequiredContract;

    /// Whether the far side has moved past what this build understands. Worth saying rather than
    /// hiding: the app keeps working against the part of the contract it knows.
    public bool IsNewerThanThisApp => ContractVersion > AgentContract.RequiredContract;

    /// Whether anything has answered yet.
    public bool IsUnknown => ContractVersion <= 0;

    // Everything below is one idea spelled once. They all say "contract 3" and they exist so that
    // the call sites read as what they need rather than as a version test.
    public bool SupportsOperationIds => SpeaksV3;
    public bool SupportsOperationQuery => SpeaksV3;
    public bool SupportsCancel => SpeaksV3;
    public bool SupportsDetach => SpeaksV3;
    public bool SupportsQueueUntilIdle => SpeaksV3;
    public bool SupportsPolicy => SpeaksV3;
    public bool SupportsCycle => SpeaksV3;
    public bool SupportsDoctor => SpeaksV3;
    public bool SupportsLogs => SpeaksV3;
    public bool SupportsHistory => SpeaksV3;
    public bool SupportsBundle => SpeaksV3;
    public bool SupportsAgentDeploy => SpeaksV3;
    /// Whether the machine can carry the setup with an identity and a lineage. A 2.x agent stores
    /// a document without either and is excluded from reconciliation rather than fought with.
    public bool SupportsSetupLineage => SpeaksV3;
    public bool SupportsConfigMeta => SpeaksV3;

    /// One line for the machine section.
    public string Summary => ContractVersion switch
    {
        <= 0 => "not read yet",
        var version when version < AgentContract.RequiredContract => $"contract {version}, older than this app needs",
        var version when version > AgentContract.RequiredContract => $"contract {version}, newer than this app knows",
        var version => $"contract {version}",
    };
}

/// Facts about the contract this build was written against.
public static class AgentContract
{
    /// What the v3 features need. An agent below this still works for everything it does have.
    public const int RequiredContract = 3;

    /// The agent version this build carries a bundle of, when it carries one.
    public const string BundledAgentVersion = "3.0.0";

    public const string ClientVersion = "1.3.0";

    /// What this client calls itself on the wire. Fixed by the contract; never a platform name.
    public const string ClientKind = "desktop";
}

File: android/app/src/main/java/com/x1f4r/legioncontrol/agent/AgentCapabilities.kt
package com.x1f4r.legioncontrol.agent

import kotlinx.serialization.Serializable

/**
 * What the agent on the far side can be asked to do.
 *
 * The contract settles this with one number. Every reply from a 3.x agent carries `"contract": 3`,
 * and a client gates every v3 feature on `contract >= 3`; absence means a 2.x or 1.x agent and the
 * old key-sniffing paths stay. That is deliberately not a capability list per verb, and it is a
 * better arrangement than one: an agent either implements contract 3 or it does not, and a partial
 * one would be a bug rather than a configuration.
 *
 * The one thing this does hold beyond the number is the verb list from `help`, and it exists for
 * exactly one command: the wake proxy, which is a proposed contract addition rather than part of
 * contract 3 proper. Nothing else consults it.
 */
data class AgentAbilities(
    /** The `contract` number from the last reply. Zero until something has answered. */
    val contract: Int = 0,
    val agentVersion: String? = null,
    /** Verb names from `help`, when it has been read. Empty otherwise, which means "do not assume". */
    val verbs: Set<String> = emptySet(),
) {
    /** Whether every v3 command and flag in this app is safe to send. */
    val speaksV3: Boolean get() = contract >= Contract.REQUIRED

    /**
     * Whether the far side is newer than this build understands.
     *
     * Worth saying rather than hiding: the app keeps working against the part of the contract it
     * knows, and the user is told that the machine has moved on.
     */
    val isNewerThanThisApp: Boolean get() = contract > Contract.REQUIRED

    /** The optional wake proxy, only ever used when `help` actually listed it. */
    val hasWakeProxy: Boolean get() = speaksV3 && "wake" in verbs

    /** One line for the machine page. */
    fun summary(): String = when {
        contract <= 0 -> "not read yet"
        !speaksV3 -> "contract $contract, older than this app needs"
        isNewerThanThisApp -> "contract $contract, newer than this app knows"
        else -> "contract $contract"
    }

    companion object {
        val unknown = AgentAbilities()

        fun of(status: AgentStatus): AgentAbilities = AgentAbilities(
            contract = status.contract ?: status.agent?.contract ?: 0,
            agentVersion = status.version,
        )

        fun of(version: AgentVersion): AgentAbilities = AgentAbilities(
            contract = version.contract ?: 0,
            agentVersion = version.agentVersion,
        )
    }
}

/** One `help` reply, read only to find out whether an optional verb exists. */
@Serializable
data class HelpReply(
    val ok: Boolean? = null,
    val contract: Int? = null,
    val agentVersion: String? = null,
    val usage: String? = null,
    val commands: List<HelpCommand>? = null,
) {
    val verbs: Set<String>
        get() = commands.orEmpty().mapNotNull { it.name?.trim()?.takeIf(String::isNotEmpty) }.toSet()
}

@Serializable
data class HelpCommand(
    val name: String? = null,
    val flags: List<String>? = null,
    val summary: String? = null,
)
